File: javascript/packages/highlighter/test/html-sink.test.ts
import { describe, test, expect, beforeAll, beforeEach } from "vitest"

import { Herb } from "@herb-tools/node-wasm"
import { Highlighter } from "../src/highlighter.js"

const PATH = "app/views/users/show.html.erb"

describe("HTMLSink", () => {
  let highlighter: Highlighter

  beforeAll(async () => {
    await Herb.load()
  })

  beforeEach(async () => {
    highlighter = new Highlighter("onedark")
    await highlighter.initialize()
  })

  test("renders a file fragment with line numbers", () => {
    const content = "<div>\n  <!-- first comment line\n  second comment line -->\n  <span>ok</span>\n</div>"
    const result = highlighter.highlightHTML(PATH, content)

    expect(result).toMatchSnapshot()
  })

  test("escapes HTML metacharacters", () => {
    const content = "<div data-x=\"a&b\"><%= a < b && b > \"c\" %>'quoted'</div>"
    const result = highlighter.highlightHTML(PATH, content)

    expect(result).toMatchSnapshot()
  })

  test("renders a plain fragment without line numbers", () => {
    const content = "<div>hello</div>\n<span>world</span>"
    const result = highlighter.highlightHTML(PATH, content, { showLineNumbers: false })

    expect(result).toMatchSnapshot()
  })

  test("renders a focus fragment with dimmed context", () => {
    const content = "<div>\n  <p>one</p>\n  <p>two</p>\n  <p>three</p>\n</div>"
    const result = highlighter.highlightHTML(PATH, content, { focusLine: 3, contextLines: 1 })

    expect(result).toMatchSnapshot()
  })

  test("renders an empty file as a single empty line", () => {
    const result = highlighter.highlightHTML(PATH, "")

    const expected = [
      "<figure class=\"herb-highlight\" data-herb-theme=\"onedark\">",
      "<figcaption class=\"herb-file-header\">app/views/users/show.html.erb</figcaption>",
      "<pre class=\"herb-code\"><code><span class=\"herb-line\" data-line=\"1\"></span></code></pre>",
      "</figure>",
    ].join("\n")

    expect(result).toBe(expected)
  })

  test("focus beyond the last line renders an empty window", () => {
    const result = highlighter.highlightHTML(PATH, "a\nb", { focusLine: 10, contextLines: 1 })

    const expected = [
      "<figure class=\"herb-highlight\" data-herb-theme=\"onedark\">",
      "<figcaption class=\"herb-file-header\">app/views/users/show.html.erb</figcaption>",
      "<pre class=\"herb-code\"><code></code></pre>",
      "</figure>",
    ].join("\n")

    expect(result).toBe(expected)
  })
})

File: rust/herb-highlighter/tests/html_sink_test.rs
mod common;

use herb_highlighter::highlighter::Highlighter;
use herb_highlighter::html_sink::HTMLRenderOptions;

const PATH: &str = "app/views/users/show.html.erb";

fn highlighter() -> Highlighter {
  Highlighter::new("onedark").expect("the theme resolves")
}

#[test]
fn renders_a_file_fragment_with_line_numbers() {
  let content = "<div>\n  <!-- first comment line\n  second comment line -->\n  <span>ok</span>\n</div>";

  insta::assert_snapshot!(highlighter().highlight_html(PATH, content, &HTMLRenderOptions::default()));
}

#[test]
fn escapes_html_metacharacters() {
  let content = "<div data-x=\"a&b\"><%= a < b && b > \"c\" %>'quoted'</div>";

  insta::assert_snapshot!(highlighter().highlight_html(PATH, content, &HTMLRenderOptions::default()));
}

#[test]
fn renders_a_plain_fragment_without_line_numbers() {
  let content = "<div>hello</div>\n<span>world</span>";

  let options = HTMLRenderOptions {
    show_line_numbers: false,
    ..Default::default()
  };

  insta::assert_snapshot!(highlighter().highlight_html(PATH, content, &options));
}

#[test]
fn renders_a_focus_fragment_with_dimmed_context() {
  let content = "<div>\n  <p>one</p>\n  <p>two</p>\n  <p>three</p>\n</div>";

  let options = HTMLRenderOptions {
    focus_line: Some(3),
    context_lines: 1,
    ..Default::default()
  };

  insta::assert_snapshot!(highlighter().highlight_html(PATH, content, &options));
}

#[test]
fn renders_an_empty_file_as_a_single_empty_line() {
  let result = highlighter().highlight_html(PATH, "", &HTMLRenderOptions::default());

  assert_eq!(
    result,
    "<figure class=\"herb-highlight\" data-herb-theme=\"onedark\">\n<figcaption class=\"herb-file-header\">app/views/users/show.html.erb</figcaption>\n<pre class=\"herb-code\"><code><span class=\"herb-line\" data-line=\"1\"></span></code></pre>\n</figure>"
  );
}

#[test]
fn focus_beyond_the_last_line_renders_an_empty_window() {
  let options = HTMLRenderOptions {
    focus_line: Some(10),
    context_lines: 1,
    ..Default::default()
  };

  let result = highlighter().highlight_html(PATH, "a\nb", &options);

  assert_eq!(
    result,
    "<figure class=\"herb-highlight\" data-herb-theme=\"onedark\">\n<figcaption class=\"herb-file-header\">app/views/users/show.html.erb</figcaption>\n<pre class=\"herb-code\"><code></code></pre>\n</figure>"
  );
}
